kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/internal/ProxyChannelStateMachine.java

```diff
@@ -0,0 +1,532 @@
+     * </p>
+     */
+    void onClientInactive() {
+        toClosed(null);
+    }
+
+    /**
+     * Notify the state machine that something exceptional and un-recoverable has happened on the upstream side.
+     * @param cause the exception that triggered the issue
+     */
+    void onServerException(Throwable cause) {
+        LOGGER.atWarn()
+                .setCause(LOGGER.isDebugEnabled() ? cause : null)
+                .addArgument(cause != null ? cause.getMessage() : "")
+                .log("Exception from the server channel: {}. Increase log level to DEBUG for stacktrace");
+        toClosed(cause);
+    }
+
+    /**
+     * Notify the state machine that something exceptional and un-recoverable has happened on the downstream side.
+     * @param cause the exception that triggered the issue
+     */
+    void onClientException(Throwable cause, boolean tlsEnabled) {
+        ApiException errorCodeEx;
+        if (cause instanceof DecoderException de
+                && de.getCause() instanceof FrameOversizedException e) {
+            var tlsHint = tlsEnabled ? "" : " or an unexpected TLS handshake";
+            LOGGER.warn(
+                    "Received over-sized frame from the client, max frame size bytes {}, received frame size bytes {} "
+                            + "(hint: are we decoding a Kafka frame, or something unexpected like an HTTP request{}?)",
+                    e.getMaxFrameSizeBytes(), e.getReceivedFrameSizeBytes(), tlsHint);
+            errorCodeEx = Errors.INVALID_REQUEST.exception();
+        }
+        else {
+            LOGGER.atWarn()
+                    .setCause(LOGGER.isDebugEnabled() ? cause : null)
+                    .addArgument(cause != null ? cause.getMessage() : "")
+                    .log("Exception from the client channel: {}. Increase log level to DEBUG for stacktrace");
+            errorCodeEx = Errors.UNKNOWN_SERVER_ERROR.exception();
+        }
+        toClosed(errorCodeEx);
+    }
+
+    private void toClientActive(
+                                @NonNull ProxyChannelState.ClientActive clientActive,
+                                @NonNull KafkaProxyFrontendHandler frontendHandler) {
+        setState(clientActive);
+        frontendHandler.inClientActive();
+    }
+
+    private void toConnecting(
+                              ProxyChannelState.Connecting connecting,
+                              @NonNull List<FilterAndInvoker> filters,
+                              VirtualCluster virtualCluster) {
+        setState(connecting);
+        backendHandler = new KafkaProxyBackendHandler(this, virtualCluster);
+        frontendHandler.inConnecting(connecting.remote(), filters, backendHandler);
+    }
+
+    private void toForwarding(Forwarding forwarding) {
+        setState(forwarding);
+        Objects.requireNonNull(frontendHandler).inForwarding();
+    }
+
+    private boolean onClientRequestBeforeForwarding(@NonNull SaslDecodePredicate dp, Object msg) {
```

Comment:
maybe add a comment on what the returned bool represents, I think false means an unexpected message type for the state.

Reply:
resolved in d6413b7d19d6aad1d1b06f787f661c393fb33ea1